cdc/sink/ddlsink/mq/ddlproducer/pulsar_ddl_producer.go

```diff
@@ -0,0 +1,247 @@
+		zap.Any("mID", mID), zap.String("topic", topic))
+
+	return nil
+}
+
+// NewPulsarProducer creates a pulsar producer
+func NewPulsarProducer(
+	ctx context.Context,
+	changefeedID model.ChangeFeedID,
+	pConfig *pulsarConfig.Config,
+	client pulsar.Client,
+) (DDLProducer, error) {
+	log.Info("Starting pulsar DDL producer ...",
+		zap.String("namespace", changefeedID.Namespace),
+		zap.String("changefeed", changefeedID.ID))
+
+	topicName := pConfig.GetDefaultTopicName()
+
+	defaultProducer, err := newProducer(pConfig, client, topicName)
+	if err != nil {
+		return nil, err
+	}
+	producers := make(map[string]pulsar.Producer)
+	producers[topicName] = defaultProducer
+	return &pulsarProducers{
+		client:           client,
+		pConfig:          pConfig,
+		producers:        producers,
+		defaultTopicName: topicName,
+		id:               changefeedID,
+	}, nil
+}
+
+// newProducer creates a pulsar producer
+// One topic is used by one producer
+func newProducer(
+	pConfig *pulsarConfig.Config,
+	client pulsar.Client,
+	topicName string,
+) (pulsar.Producer, error) {
+	po := pulsar.ProducerOptions{
+		Topic: topicName,
+	}
+	if pConfig.BatchingMaxMessages > 0 {
+		po.BatchingMaxMessages = pConfig.BatchingMaxMessages
+	}
+	if pConfig.BatchingMaxPublishDelay > 0 {
+		po.BatchingMaxPublishDelay = pConfig.BatchingMaxPublishDelay
+	}
+	if pConfig.CompressionType > 0 {
+		po.CompressionType = pConfig.CompressionType
+		po.CompressionLevel = pulsar.Default
+	}
+	if pConfig.SendTimeout > 0 {
+		po.SendTimeout = pConfig.SendTimeout
+	}
+
+	producer, err := client.CreateProducer(po)
+	if err != nil {
+		return nil, err
+	}
+
+	log.L().Info("create pulsar producer success",
+		zap.String("topic:", topicName))
+
+	return producer, nil
+}
+
+// GetProducerByTopic get producer by topicName
+func (p *pulsarProducers) GetProducerByTopic(topicName string) (producer pulsar.Producer, err error) {
+	p.producersMutex.RLock()
+	producer, ok := p.producers[topicName]
+	p.producersMutex.RUnlock()
+	if !ok { // create a new producer for the topicName
+		p.producersMutex.Lock()
+		defer p.producersMutex.Unlock()
+
+		producer, ok = p.producers[topicName]
+		if ok {
+			return producer, nil
+		}
+
+		producer, err = newProducer(p.pConfig, p.client, topicName)
+		if err != nil {
+			return nil, err
+		}
+		p.producers[topicName] = producer
+	}
+
+	return producer, nil
+}
+
+// Close close all producers
+func (p *pulsarProducers) Close() {
+	for topic, producer := range p.producers {
+		producer.Close()
+		p.closeProducersMapByTopic(topic)
+	}
+	p.client.Close()
+}
+
+// Flush waits for all the messages in the async producer to be sent to Pulsar.
+// Notice: this method is not thread-safe.
+// Do not try to call AsyncSendMessage and Flush functions in different threads,
+// otherwise Flush will not work as expected. It may never finish or flush the wrong message.
+// Because inflight will be modified by mistake.
+func (p *pulsarProducers) Flush(ctx context.Context) error {
```

Comment:
It seems that there is no `AsyncSendMessage` method in this file.

Reply:
I will remove it